sql/src/main/java/io/crate/action/job/ContextPreparer.java

```diff
@@ -88,65 +97,207 @@ public ContextPreparer(MapSideDataCollectOperation collectOperation,
+            }
+        }
+        postPrepare(contextBuilder, preparerContext);
+
+        return directResponseFutures;
     }
 
-    @SuppressWarnings("unchecked")
-    public <T extends ExecutionSubContext> T prepare(UUID jobId, ExecutionPhase executionPhase, RowReceiver rowReceiver) {
-        PreparerContext preparerContext = new PreparerContext(jobId, null, rowReceiver);
-        return (T) innerPreparer.process(executionPhase, preparerContext);
+    @Nullable
+    public ExecutionSubContext prepareOnHandler(Iterable<? extends NodeOperation> nodeOperations,
+                                                          JobExecutionContext.Builder contextBuilder,
+                                                          PreparerContext preparerContext,
+                                                          ExecutionPhase handlerMergePhase,
+                                                          RowReceiver handlerPhaseRowReceiver) {
+        preparerContext.rowDownstreamFactory = rowDownstreamFactory;
+        processDownstreamExecutionPhaseIds(nodeOperations, preparerContext);
+
+        // register handler phase row receiver
+        preparerContext.rowReceiver(handlerMergePhase.executionPhaseId(), handlerPhaseRowReceiver);
+        // build handler context, must be done first because it's downstream is already known
+        // and it is needed as a row receiver by others
+        ExecutionSubContext finalLocalMergeContext = innerPreparer.process(handlerMergePhase, preparerContext);
+        if (finalLocalMergeContext != null) {
+            contextBuilder.addSubContext(handlerMergePhase.executionPhaseId(), finalLocalMergeContext);
+        }
+
+        List<NodeOperation> reversedNodeOperations = Lists.reverse(Lists.newArrayList(nodeOperations));
+        for (NodeOperation nodeOperation : reversedNodeOperations) {
+            ExecutionSubContext subContext = innerPreparer.process(nodeOperation.executionPhase(), preparerContext);
+            if (subContext != null) {
+                contextBuilder.addSubContext(nodeOperation.executionPhase().executionPhaseId(), subContext);
+            }
+        }
+        postPrepare(contextBuilder, preparerContext);
+        return finalLocalMergeContext;
     }
 
-    private static class PreparerContext {
+
+    /**
+     * Build all contexts which could not build in first iteration through to missing downstreams
+     */
+    private void postPrepare(JobExecutionContext.Builder contextBuilder,
+                             PreparerContext preparerContext) {
+        while (!preparerContext.executionPhasesToProcess.isEmpty()) {
+            Iterator<ExecutionPhase> it = preparerContext.executionPhasesToProcess.iterator();
+            while (it.hasNext()) {
+                ExecutionPhase executionPhase = it.next();
+                it.remove();
+                ExecutionSubContext subContext = innerPreparer.process(executionPhase, preparerContext);
```

Comment:
process + nullCheck + addSubContext could probably be moved into a static helper method. It is done multiple times